prometheus/promhttp/http.go

```diff
@@ -165,20 +188,32 @@ func HandlerForTransactional(reg prometheus.TransactionalGatherer, opts HandlerO
 		} else {
 			contentType = expfmt.Negotiate(req.Header)
 		}
-		header := rsp.Header()
-		header.Set(contentTypeHeader, string(contentType))
+		rsp.Header().Set(contentTypeHeader, string(contentType))
 
-		w := io.Writer(rsp)
-		if !opts.DisableCompression && gzipAccepted(req.Header) {
-			header.Set(contentEncodingHeader, "gzip")
-			gz := gzipPool.Get().(*gzip.Writer)
-			defer gzipPool.Put(gz)
+		w, encodingHeader, closeWriter, err := NegotiateEncodingWriter(req, rsp, opts.DisableCompression, compressions)
 
-			gz.Reset(w)
-			defer gz.Close()
+		if closeWriter != nil {
+			defer func() {
+				err := closeWriter()
+				if err != nil {
+					if opts.ErrorLog != nil {
+						opts.ErrorLog.Println("error closing writer:", err)
+					}
+				}
+			}()
+		}
+		if err != nil {
+			if opts.ErrorLog != nil {
+				opts.ErrorLog.Println("error getting writer", err)
+			}
+			// Since the writer received from NegotiateEncodingWriter will be nil, in case there's an error, we set it here
+			w = io.Writer(rsp)
+		}
 
-			w = gz
+		if encodingHeader == "" {
```

Comment:
After err check we expect this to be always valid, can we ensure in func?

Reply:
Good idea, I moved it into the err check.